### DISCUSSION TITLE
Suitability of xgcm for interpolation of point cloud data on sub-grid scale

### DISCUSSION THREAD

**Discussion by VeckoTheGecko** [xgcm/xgcm#658] [Q&A]
Hello! I am a core developer for [Parcels](https://github.com/OceanParcels/Parcels), a Lagrangian simulation toolbox. Our core functionality is releasing particles to be advected on GCM output according to user defined behaviour.

We're looking to migrate our internals to xarray ([issue](https://github.com/OceanParcels/Parcels/issues/1796)). 

I was very excited to find xgcm for its ability to interpret Arakawa grids, and automatically parse metadata from NetCDF datasets. Initially the presence of  `Grid.interp()` gave me hope that we would be able to heavily rely on this package, however looking deeper at `interp()` I since found that its specialised for interpolating dataarrays **onto** the grid and not for arbitrary interpolation using the context of the grid.

Also looking at the public API, I don't see how I could plug into xgcm to extract the grid context for use in custom interpolation routines. Any suggestions? Or is xgcm very specialised to the world of finite volume analysis to the point we're better off writing our own code?

cc @erikvansebille

**Comment by dcherian** [comment]
What kind of interpolation do you want? From gridded field to trajectory?

**Reply by VeckoTheGecko** [reply]
Yes, from the gridded field onto points (which would be used to compute the trajectory)

**Reply by dcherian** [reply]
Xarray's interp implements this "vectorized interpolation", but not always the most performant. I am planning on fixing that for `method="linear"` and `method="interp"` though. What interpolation methods are you interested in?

**Reply by VeckoTheGecko** [reply]
We're mainly interested in linear and nearest neighbour interpolation... also on arbitrary curvilinear grids (which looks to be an open issue https://github.com/pydata/xarray/issues/2281#issuecomment-404744836).

But in the case of a rectilinear C grid, I guess it would be a matter of:

- Ensuring the dataset dimension naming conforms to some standard (e.g., Comodo like here)
- doing `ds["V"].interp(XC = ..., YG = ...)`

and bypassing xgcm entirely. If the grid file is separate, then that is a separate can of worms if going via xarray.

I was just hoping that we could piggyback a bit more off of xgcm, but I don't know if thats possible.
